common/raw_hashtable.h

```diff
@@ -0,0 +1,1206 @@
+  static_assert((SmallSize % MaxGroupSize) == 0,
+                "Small size must be a multiple of the max group size supported "
+                "so that we can allocate a whole number of groups.");
+  // Implied by the max asserts above.
+  static_assert(SmallSize >= GroupSize);
+  static_assert((SmallSize % GroupSize) == 0);
+
+  static_assert(SmallSize >= alignof(StorageEntry<KeyT, ValueT>),
+                "Requested a small size that would require padding between "
+                "metadata bytes and correctly aligned key and value types. "
+                "Either a larger small size or a zero small size and heap "
+                "allocation are required for this key and value type.");
+
+  // A concrete implementation of storage for the provided key type, value type,
+  // and small size.
+  struct SmallStorage : Storage {
+    alignas(BaseT::Alignment) uint8_t metadata[SmallSize];
+    mutable StorageEntry<KeyT, ValueT> entries[SmallSize];
+  };
+
+  auto small_storage() const -> RawHashtable::Storage* {
+    return &small_storage_;
+  }
+
+  mutable SmallStorage small_storage_;
+};
+
+// A template specialization for when no small size optimization buffer is
+// desired.
+template <typename InputBaseT>
+class TableImpl<InputBaseT, 0> : public InputBaseT {
+ protected:
+  using BaseT = InputBaseT;
+
+  TableImpl() : BaseT(0, nullptr) {}
+  TableImpl(const TableImpl& arg) : TableImpl() { this->CopyFrom(arg); }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(const TableImpl<BaseT, OtherMinSmallSize>& arg)
+      : TableImpl() {
+    this->CopyFrom(arg);
+  }
+  TableImpl(TableImpl&& arg) noexcept : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+  template <ssize_t OtherMinSmallSize>
+  explicit TableImpl(TableImpl<BaseT, OtherMinSmallSize>&& arg) : TableImpl() {
+    this->MoveFrom(std::move(arg));
+  }
+
+  auto ResetImpl() -> void;
+};
+
+// Computes a seed that provides a small amount of entropy from ASLR where
+// possible with minimal cost. The priority is speed, and this computes the
+// entropy in a way that doesn't require loading from memory, merely accessing
+// entropy already available without accessing memory.
+inline auto ComputeSeed() -> uint64_t {
+  // A global variable whose address is used as a seed. This allows ASLR to
+  // introduce some variation in hashtable ordering when enabled via the code
+  // model for globals.
+  extern volatile std::byte global_addr_seed;
+
+  return reinterpret_cast<uint64_t>(&global_addr_seed);
+}
+
+inline auto ComputeProbeMaskFromSize(ssize_t size) -> size_t {
+  CARBON_DCHECK(llvm::isPowerOf2_64(size))
+      << "Size must be a power of two for a hashed buffer!";
+  // The probe mask needs to mask down to keep the index within
+  // `size`. Since `size` is a power of two, this is equivalent to
+  // `size - 1`. We also mask off the low bits while here to match the size of
+  // the groups of entries.
```

Comment:
"The probe mask needs to mask down" and "equivalent to `size -1`" are awkward and unclear. Perhaps:

```suggestion
  // Since `size` is a power of two, we can make sure the probes are less
  // than `size` by making the mask `size - 1`. We also mask off the low
  // bits so the probes are a multiple of the size of the groups of entries.
```